### PR TITLE
Feature/217-Neural-based-dialogue-policies

### PR DESCRIPTION
**Pull request by NoB0**
Add neural policies: DQN and A2C.

### CODE REVIEW

File: tests/dialogue_manager/test_dialogue_policy.py [hunk 1]
```diff
@@ -206,9 +210,12 @@ def test_next_action_accept_recommendation(
...
+    "moviebot.dialogue_manager.dialogue_policy.rb_dialogue_policy.set",
+    mock.MagicMock(wraps=list),
 )
-def test__generate_examples(policy: DialoguePolicy, results, slot, expected):
+def test__generate_examples(
+    policy: RuleBasedDialoguePolicy, results, slot, expected
+):
     random.seed(42)
     examples = policy._generate_examples(results, slot)
     assert examples == expected
```
Comment (WerLaj) [inline tests/dialogue_manager/test_dialogue_policy.py:221]: Would it be possible to have some tests for these two new policies?

**Review by WerLaj — COMMENTED** [review]
Some clarifications are requested

File: moviebot/dialogue_manager/dialogue_policy/neural_dialogue_policy.py [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+            )
+        else:
+            user_intents_tensor = torch.tensor(
+                cls.user_label_encoder.transform(
+                    [list(map(lambda x: x.value.label, user_intents))]
+                )[0],
+                dtype=torch.float,
+            )
+
+        if len(agent_intents) == 0:
```
[the remark this answers is not in the document]
Comment (NoB0) [reply moviebot/dialogue_manager/dialogue_policy/neural_dialogue_policy.py:169]: Good point, I created a new class method.

File: moviebot/dialogue_manager/dialogue_policy/neural_dialogue_policy.py [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+
+        Args:
+            dialogue_state: The dialogue state.
+
+        Returns:
+            The input vector.
+        """
+        dialogue_state_tensor = torch.tensor(
+            [
+                dialogue_state.is_beginning,
```
[the remark this answers is not in the document]
Comment (NoB0) [reply moviebot/dialogue_manager/dialogue_policy/neural_dialogue_policy.py:122]: I added a short explanation in the docstrings, please check if it helps to better understand how the dialogue state is represented.

File: moviebot/dialogue_manager/dialogue_policy/a2c_dialogue_policy.py [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+        Args:
+            state: Representation of dialogue state as a vector.
+
+        Returns:
+            The selected action id, its log probability, the state value, and
+            the entropy.
+        """
+        state_value, actions_log_prob = self.forward(state)
+        action_pd = torch.distributions.Categorical(logits=actions_log_prob)
+        action = action_pd.sample()
```
[the remark this answers is not in the document]
Comment (NoB0) [reply moviebot/dialogue_manager/dialogue_policy/a2c_dialogue_policy.py:93]: It is a tensor.

File: moviebot/dialogue_manager/dialogue_policy/a2c_dialogue_policy.py [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+
+        Args:
+            state: Representation of dialogue state as a vector.
+
+        Returns:
+            The selected action id, its log probability, the state value, and
+            the entropy.
+        """
+        state_value, actions_log_prob = self.forward(state)
+        action_pd = torch.distributions.Categorical(logits=actions_log_prob)
```
[the remark this answers is not in the document]
Comment (NoB0) [reply moviebot/dialogue_manager/dialogue_policy/a2c_dialogue_policy.py:92]: It is a categorical probability distribution, each category (in this case action) has a certain probability based on the logits from the actor.

File: tests/dialogue_manager/test_dialogue_policy.py (hunk 1 above)
Comment (NoB0) [reply tests/dialogue_manager/test_dialogue_policy.py:221]: I created a follow-up issue #229 for that.

File: moviebot/dialogue_manager/dialogue_policy/neural_dialogue_policy.py [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+import torch
+from sklearn.preprocessing import MultiLabelBinarizer
+
+from moviebot.core.intents.agent_intents import AgentIntents
+from moviebot.core.intents.user_intents import UserIntents
+from moviebot.dialogue_manager.dialogue_state import DialogueState
+
+
+class NeuralDialoguePolicy(torch.nn.Module):
+    user_label_encoder = MultiLabelBinarizer().fit(
```
[the remark this answers is not in the document]
Comment (WerLaj) [reply moviebot/dialogue_manager/dialogue_policy/neural_dialogue_policy.py:17]: Got it, thanks

File: moviebot/dialogue_manager/dialogue_policy/neural_dialogue_policy.py [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+                dialogue_state.agent_made_offer,
+                dialogue_state.agent_offer_no_results,
+                dialogue_state.at_terminal_state,
+            ],
+            dtype=torch.float,
+        )
+        return dialogue_state_tensor
+
+    @classmethod
+    def _encode_intents(cls, intents: List[Any], label_encoder) -> torch.Tensor:
```
Comment (WerLaj) [inline moviebot/dialogue_manager/dialogue_policy/neural_dialogue_policy.py:141]: Missing type for label_encoder

**Review by WerLaj — APPROVED** [review]
LGTM, with one comment to address. Thank you for clarifications.

File: moviebot/dialogue_manager/dialogue_policy/neural_dialogue_policy.py (hunk 7 above)
Comment (NoB0) [reply moviebot/dialogue_manager/dialogue_policy/neural_dialogue_policy.py:141]: Done.
